testing/framework/TestCommon.py

```diff
@@ -292,46 +295,47 @@ def must_be_writable(self, *files):
             print("Unwritable files: `%s'" % "', `".join(unwritable))
         self.fail_test(missing + unwritable)
 
-    def must_contain(self, file, required, mode='rb', find=None):
-        """Ensures specified file contains the required text.
-
-        Args:
-            file (string): name of file to search in.
-            required (string): text to search for. For the default
-              find function, type must match the return type from
-              reading the file; current implementation will convert.
-            mode (string): file open mode.
-            find (func): optional custom search routine. Must take the
-              form "find(output, line)" non-negative integer on success
-              and None, False, or -1, on failure.
+    def must_contain(self, fname, required, mode='rb', find=None):
+        """Ensures a file contains the required text.
 
         Calling test exits FAILED if search result is false
+
+        :param str fname: name of file to search in.
+        :param required: text to search for. For the default
+            find function, type must match the return type from
+            reading the file; current implementation will convert.
+        :type required: str or bytes
+        :param str mode: file open mode.
+        :param callable find: (optional) custom search routine. Must take the
+            form "find(output, line)", returning a true value on success
+            and a false value on failure.
         """
+        file_contents = self.read(fname, mode)
         if 'b' in mode:
-            # Python 3: reading a file in binary mode returns a 
+            # Python 3: reading a file in binary mode returns a
```

Comment:
to_str() doesn't always work. If you have an actual binary file and you're looking for a series of bytes. this will break?

Reply:
Hmmm.  The testing does entirely too much stuff in binary mode, but it's too expensive to "fix" (I wasted some time going down that road at one point(. This was just trying to normalize where things are converted.

Reply:
So swinging back to this... this location, it will call `TestCommon:contains` (line 317).  The two arguments, `file_contents` and `required` have to be of the same type. I've switched this so instead of `required` being converted to bytes we're converting `file_contents` to string.  Are you saying that's somehow less reliable?

Reply:
yes.

Reply:
if it's unconvertible binary data, then the "in" membership operator that gets used is unlikely to be useful anyway... does this need further rethinking?